main.go

```diff
@@ -200,6 +166,58 @@ func initRequestHandlers(stopChannel <-chan struct{}, waitGroup *sync.WaitGroup)
 	log.Debug().Msg("Adding auth middleware...")
 	authMiddleware := auth.NewAuthMiddleware(*config.Auth)
 
+	initRoutes(router, authMiddleware, githubEventHandler, bitbucketEventHandler, slackEventHandler, pubsubEventHandler, estafetteEventHandler, estafetteAPIHandler)
+
+	// instantiate servers instead of using router.Run in order to handle graceful shutdown
+	log.Debug().Msg("Starting server...")
+	srv := &http.Server{
+		Addr:        *apiAddress,
+		Handler:     router,
+		ReadTimeout: 30 * time.Second,
+		//WriteTimeout:   30 * time.Second,
+		MaxHeaderBytes: 1 << 20,
+	}
+
+	// watch for configmap changes
+	foundation.WatchForFileChanges(*configFilePath, func(event fsnotify.Event) {
+		log.Info().Msgf("Configmap at %v was updated, refreshing instances...", *configFilePath)
+
+		config, err = configReader.ReadConfigFromFile(*configFilePath, true)
+		if err != nil {
+			log.Fatal().Err(err).Msg("Failed reading configuration")
+		}
+		encryptedConfig, err = configReader.ReadConfigFromFile(*configFilePath, false)
+		if err != nil {
+			log.Fatal().Err(err).Msg("Failed reading configuration without decrypting")
+		}
+
+		// refresh instances
+		githubEventHandler, bitbucketEventHandler, slackEventHandler, pubsubEventHandler, estafetteEventHandler, estafetteAPIHandler = getInstances(secretHelper, config, encryptedConfig)
+
+		initRoutes(router, authMiddleware, githubEventHandler, bitbucketEventHandler, slackEventHandler, pubsubEventHandler, estafetteEventHandler, estafetteAPIHandler)
+	})
+
+	// watch for service account key file changes
+	foundation.WatchForFileChanges(os.Getenv("GOOGLE_APPLICATION_CREDENTIALS"), func(event fsnotify.Event) {
+		log.Info().Msg("Service account key file was updated, refreshing instances...")
+
+		// refresh instances
+		githubEventHandler, bitbucketEventHandler, slackEventHandler, pubsubEventHandler, estafetteEventHandler, estafetteAPIHandler = getInstances(secretHelper, config, encryptedConfig)
+
+		initRoutes(router, authMiddleware, githubEventHandler, bitbucketEventHandler, slackEventHandler, pubsubEventHandler, estafetteEventHandler, estafetteAPIHandler)
```

Comment:
I'm not sure whether you can set routes a second time and whether they'll overwrite the initial ones. Did find https://github.com/gin-gonic/gin/issues/776 but it's doesn't help a lot.

Reply:
I tried returning the instances as pointers, but then you can't set up routes. It complains the instance don't implement those methods.